### PR TITLE
Allow root admin to deploy in VPCs in child domains

### PR DESCRIPTION
**Pull request by DaanHoogland**
### Description

This PR...

Fixes: #6754 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
the component tests test_acl_sharednetwork_deployVM-impersonation, test_shared_networks, test_user_private_gateway and test_user_shared_network are all running in my test env but keep failing in travis job 8.

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 1]
```diff
@@ -1665,39 +1666,49 @@ public void checkCapabilityForProvider(Set<Provider> providers, Service service,
...
-        if (network.getGuestType() != GuestType.Shared || network.getAclType() == ACLType.Account) {
-            AccountVO networkOwner = _accountDao.findById(network.getAccountId());
-            if (networkOwner == null)
-                throw new PermissionDeniedException("Unable to use network with id= " + ((NetworkVO)network).getUuid() +
-                    ", network does not have an owner");
-            if (!Account.Type.PROJECT.equals(caller.getType()) && Account.Type.PROJECT.equals(networkOwner.getType())) {
-                checkProjectNetworkPermissions(caller, networkOwner, network);
+    public final void checkNetworkPermissions(Account caller, Network network) {
+        if (_accountMgr.isRootAdmin(caller.getAccountId())) {
+            if (s_logger.isDebugEnabled()) {
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/network/NetworkModelImpl.java:1671]: @DaanHoogland 
this looks more clean
```
        if (_accountMgr.isRootAdmin(caller.getAccountId())) {
            return;
        }
```

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:1671]: I don't agree. An extra exit point from a method is never good. I'll see if I can refactor it.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing.

**Comment by DaanHoogland** [comment]
I doubt this should be implemented. It breaks some component tests that make clear that it was working as designed. see `test_acl_sharednetwork.py`

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java (hunk 1 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:1671]: ok @DaanHoogland

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java [hunk 2]
```diff
@@ -1665,39 +1666,49 @@ public void checkCapabilityForProvider(Set<Provider> providers, Service service,
...
+                s_logger.debug("root admin is permitted to do stuff on every network");
+            }
+        } else {
+            if (network == null) {
+                throw new CloudRuntimeException("cannot check permissions on (Network) <null>");
+            }
+            s_logger.info(String.format("Checking permission for account %s (%s) on network %s (%s)", caller.getAccountName(), caller.getUuid(), network.getName(), network.getUuid()));
+            if (network.getGuestType() != GuestType.Shared || network.getAclType() == ACLType.Account) {
+                checkSharedNetworkPermissions(caller, network);
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/network/NetworkModelImpl.java:1680]: @DaanHoogland 
the name of the methods seem misunderstanding.
`checkSharedNetworkPermissions` includes the check on L2/Isolated networks and Account-level share networks. in summary, the owner of the network is an account.
`checkIsolatedNetworkPermissions` includes the check on Domain-level share networks.

**Review by weizhouapache — APPROVED** [review]
code lgtm
left a comment, @DaanHoogland

File: server/src/main/java/com/cloud/network/NetworkModelImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/network/NetworkModelImpl.java:1680]: how about 
`checkSharedNetworkPermissions` ==>> `checkSharedOrAccountLevelPermissions` and
`checkIsolatedNetworkPermissions` ==>> `checkNetworkAvailableForDomain` ?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
> Trillian test result (tid-5570) Environment: vmware-65u2 (x2), Advanced Networking with Mgmt server 7 Total time taken: 43643 seconds Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6832-t5570-vmware-65u2.zip Smoke tests completed. 104 look OK, 1 have errors, 0 did not run Only failed and skipped tests results shown below:
> Test 	Result 	Time (s) 	Test File
> test_02_create_template_with_checksum_sha1 	`Error` 	5.15 	test_templates.py
> test_03_create_template_with_checksum_sha256 	`Error` 	5.15 	test_templates.py
> test_04_create_template_with_checksum_md5 	`Error` 	5.14 	test_templates.py

see #7001 for a fix for these errors. they have to do with a template fix (in the setting pw script on vmware)

**Pull request merged**